### PR TITLE
Linux Artist Mode: Implement IToolReport

### PR DESCRIPTION
**Pull request by gonX**
This allows programs like Krita to track per-stylus settings

This supersedes d55af3f09f598435a83c14407266e9cf29b75658 from #1984

(please test, the changes were simply rebased to the current codebase)

### CODE REVIEW

**Review by X9VoiD — APPROVED** [review]

File: OpenTabletDriver.Desktop/Interop/Input/Absolute/EvdevVirtualTablet.cs [hunk 1]
```diff
@@ -141,10 +153,16 @@ public void Reset()
...
+            toolID = toolSerial = 0;
         }
 
         public void Flush()
         {
+            if (toolID > 0)
+                Device.Write(EventType.EV_ABS, EventCode.ABS_MISC, toolID);
+            if (toolSerial > 0)
+                Device.Write(EventType.EV_MSC, EventCode.MSC_SERIAL, toolSerial);
+
```
[the remark this answers is not in the document]
Comment (gonX) [reply OpenTabletDriver.Desktop/Interop/Input/Absolute/EvdevVirtualTablet.cs:165]: It exists so we don't emit a tool report on the Flush following a Reset.

I wrote the code a while back but I believe it exists to match the behavior of the in-kernel driver, but I could double-check to verify.

**Review by InfinityGhost — DISMISSED** [review]

**Comment by InfinityGhost** [comment]
Resolve merge conflicts

**Comment by gonX** [comment]
Sorry for delay, rebased now.

**Review by X9VoiD — APPROVED** [review]

**Comment by gonX** [comment]
Some personal notes for finalizing this PR (aside from it being recently rebased):

PTH-851 in-out motion on `wacom` driver:
```
$ libinput debug-events /dev/input/event20
-event20  DEVICE_ADDED            Wacom Intuos Pro L Pen            seat0 default group1  cap:T  size 325x203mm left
 event20  TABLET_TOOL_PROXIMITY   +0.000s		198.66*/114.06*	tilt: 6.03*/2.01*	distance: 0.86*	pen      (0x5c8006da, id 0x100802) proximity-in 	axes:dpt	btn:SS2
 event20  TABLET_TOOL_AXIS        +0.005s		197.94*/113.03*	tilt: 3.02*/1.01	distance: 1.00*
 event20  TABLET_TOOL_AXIS        +0.010s		197.28*/111.90*	tilt: 1.01*/1.01	distance: 1.00
 event20  TABLET_TOOL_PROXIMITY   +0.060s		197.28/111.90	tilt: 1.01/1.01	distance: 1.00	pen      (0x5c8006da, id 0x100802) proximity-out 
 ```
```
$ evtest /dev/input/event20
Input driver version is 1.0.1
Input device ID: bus 0x3 vendor 0x56a product 0x317 version 0x110
Input device name: "Wacom Intuos Pro L Pen"
Supported events:
  Event type 0 (EV_SYN)
  Event type 1 (EV_KEY)
    Event code 272 (BTN_LEFT)
    Event code 273 (BTN_RIGHT)
    Event code 274 (BTN_MIDDLE)
    Event code 275 (BTN_SIDE)
    Event code 276 (BTN_EXTRA)
    Event code 320 (BTN_TOOL_PEN)
    Event code 321 (BTN_TOOL_RUBBER)
    Event code 322 (BTN_TOOL_BRUSH)
    Event code 323 (BTN_TOOL_PENCIL)
    Event code 324 (BTN_TOOL_AIRBRUSH)
    Event code 326 (BTN_TOOL_MOUSE)
    Event code 327 (BTN_TOOL_LENS)
    Event code 330 (BTN_TOUCH)
    Event code 331 (BTN_STYLUS)
    Event code 332 (BTN_STYLUS2)
  Event type 2 (EV_REL)
    Event code 8 (REL_WHEEL)
  Event type 3 (EV_ABS)
    Event code 0 (ABS_X)
      Value      0
      Min        0
      Max    65024
      Fuzz       4
      Resolution     200
    Event code 1 (ABS_Y)
      Value      0
      Min        0
      Max    40640
      Fuzz       4
      Resolution     200
    Event code 2 (ABS_Z)
      Value      0
      Min     -900
      Max      899
      Resolution     287
    Event code 5 (ABS_RZ)
      Value      0
      Min     -900
      Max      899
      Resolution     287
    Event code 6 (ABS_THROTTLE)
      Value      0
      Min    -1023
      Max     1023
    Event code 8 (ABS_WHEEL)
      Value      0
      Min        0
      Max     1023
    Event code 24 (ABS_PRESSURE)
      Value      0
      Min        0
      Max     2047
    Event code 25 (ABS_DISTANCE)
      Value      0
      Min        0
      Max       63
      Fuzz       1
    Event code 26 (ABS_TILT_X)
      Value      0
      Min      -64
      Max       63
      Fuzz       1
      Resolution      57
    Event code 27 (ABS_TILT_Y)
      Value      0
      Min      -64
      Max       63
      Fuzz       1
      Resolution      57
    Event code 40 (ABS_MISC)
      Value      0
      Min        0
      Max        0
  Event type 4 (EV_MSC)
    Event code 0 (MSC_SERIAL)
Properties:
  Property type 0 (INPUT_PROP_POINTER)
Testing ... (interrupt to exit)
Event: time 1698196090.501512, type 3 (EV_ABS), code 0 (ABS_X), value 39732
Event: time 1698196090.501512, type 3 (EV_ABS), code 1 (ABS_Y), value 22812
Event: time 1698196090.501512, type 3 (EV_ABS), code 25 (ABS_DISTANCE), value 54
Event: time 1698196090.501512, type 3 (EV_ABS), code 26 (ABS_TILT_X), value 6
Event: time 1698196090.501512, type 3 (EV_ABS), code 27 (ABS_TILT_Y), value 2
Event: time 1698196090.501512, type 3 (EV_ABS), code 40 (ABS_MISC), value 1050626
Event: time 1698196090.501512, type 1 (EV_KEY), code 320 (BTN_TOOL_PEN), value 1
Event: time 1698196090.501512, type 4 (EV_MSC), code 0 (MSC_SERIAL), value 1551894234
Event: time 1698196090.501512, -------------- SYN_REPORT ------------
Event: time 1698196090.506513, type 3 (EV_ABS), code 0 (ABS_X), value 39587
Event: time 1698196090.506513, type 3 (EV_ABS), code 1 (ABS_Y), value 22606
Event: time 1698196090.506513, type 3 (EV_ABS), code 25 (ABS_DISTANCE), value 63
Event: time 1698196090.506513, type 3 (EV_ABS), code 26 (ABS_TILT_X), value 3
Event: time 1698196090.506513, type 3 (EV_ABS), code 27 (ABS_TILT_Y), value 1
Event: time 1698196090.506513, type 4 (EV_MSC), code 0 (MSC_SERIAL), value 1551894234
Event: time 1698196090.506513, -------------- SYN_REPORT ------------
Event: time 1698196090.511513, type 3 (EV_ABS), code 0 (ABS_X), value 39455
Event: time 1698196090.511513, type 3 (EV_ABS), code 1 (ABS_Y), value 22380
Event: time 1698196090.511513, type 3 (EV_ABS), code 26 (ABS_TILT_X), value 1
Event: time 1698196090.511513, type 4 (EV_MSC), code 0 (MSC_SERIAL), value 1551894234
Event: time 1698196090.511513, -------------- SYN_REPORT ------------
Event: time 1698196090.645512, type 3 (EV_ABS), code 0 (ABS_X), value 0
Event: time 1698196090.645512, type 3 (EV_ABS), code 1 (ABS_Y), value 0
Event: time 1698196090.645512, type 3 (EV_ABS), code 25 (ABS_DISTANCE), value 0
Event: time 1698196090.645512, type 3 (EV_ABS), code 26 (ABS_TILT_X), value 0
Event: time 1698196090.645512, type 3 (EV_ABS), code 27 (ABS_TILT_Y), value 0
Event: time 1698196090.645512, type 1 (EV_KEY), code 320 (BTN_TOOL_PEN), value 0
Event: time 1698196090.645512, type 3 (EV_ABS), code 40 (ABS_MISC), value 0
Event: time 1698196090.645512, type 4 (EV_MSC), code 0 (MSC_SERIAL), value 1551894234
Event: time 1698196090.645512, -------------- SYN_REPORT ------------
^C
```

Summed up, wacom driver does the following:

- On entry
-- Set _all_ values (x/y/distance/misc/serial/tool)
- Per-report
-- `MSC_SERIAL` and `ABS_MISC` set to the same value (continuous ABS_MISC does not appear with `evtest` because unchanging absolute values are not reported)
- On exit:
-- Set everything, except for MSC_SERIAL, to 0. Internal serial ID is likely unset after this last flush.

I will undraft the PR once the above features are matched.

**Comment by gonX** [comment]
TL;DR: tool reports work now, at the cost of lift off height

OTD now reports:
```
Input driver version is 1.0.1
Input device ID: bus 0x0 vendor 0x0 product 0x0 version 0x0
Input device name: "OpenTabletDriver Virtual Artist Tablet"
Supported events:
  Event type 0 (EV_SYN)
  Event type 1 (EV_KEY)
    Event code 320 (BTN_TOOL_PEN)
    Event code 321 (BTN_TOOL_RUBBER)
    Event code 329 (BTN_STYLUS3)
    Event code 330 (BTN_TOUCH)
    Event code 331 (BTN_STYLUS)
    Event code 332 (BTN_STYLUS2)
  Event type 3 (EV_ABS)
    Event code 0 (ABS_X)
      Value      0
      Min        0
      Max   4480000
      Resolution  100000
    Event code 1 (ABS_Y)
      Value      0
      Min        0
      Max   1440000
      Resolution  100000
    Event code 24 (ABS_PRESSURE)
      Value      0
      Min        0
      Max    65535
    Event code 26 (ABS_TILT_X)
      Value      0
      Min      -64
      Max       63
      Resolution      57
    Event code 27 (ABS_TILT_Y)
      Value      0
      Min      -64
      Max       63
      Resolution      57
    Event code 40 (ABS_MISC)
      Value      0
      Min        0
      Max        0
  Event type 4 (EV_MSC)
    Event code 0 (MSC_SERIAL)
Properties:
  Property type 0 (INPUT_PROP_POINTER)
  Property type 1 (INPUT_PROP_DIRECT)
Testing ... (interrupt to exit)
Event: time 1698250985.865976, type 3 (EV_ABS), code 40 (ABS_MISC), value 67586
Event: time 1698250985.865976, type 4 (EV_MSC), code 0 (MSC_SERIAL), value 1551894234
Event: time 1698250985.865976, -------------- SYN_REPORT ------------
Event: time 1698250985.871129, type 3 (EV_ABS), code 26 (ABS_TILT_X), value 8
Event: time 1698250985.871129, type 3 (EV_ABS), code 27 (ABS_TILT_Y), value -12
Event: time 1698250985.871129, type 3 (EV_ABS), code 24 (ABS_PRESSURE), value 13638
Event: time 1698250985.871129, type 1 (EV_KEY), code 320 (BTN_TOOL_PEN), value 1
Event: time 1698250985.871129, type 3 (EV_ABS), code 0 (ABS_X), value 1253681
Event: time 1698250985.871129, type 3 (EV_ABS), code 1 (ABS_Y), value 605728
Event: time 1698250985.871129, type 4 (EV_MSC), code 0 (MSC_SERIAL), value 1551894234
Event: time 1698250985.871129, -------------- SYN_REPORT ------------
Event: time 1698250985.875094, type 3 (EV_ABS), code 26 (ABS_TILT_X), value 6
Event: time 1698250985.875094, type 3 (EV_ABS), code 27 (ABS_TILT_Y), value -14
Event: time 1698250985.875094, type 3 (EV_ABS), code 24 (ABS_PRESSURE), value 0
Event: time 1698250985.875094, type 3 (EV_ABS), code 0 (ABS_X), value 1258405
Event: time 1698250985.875094, type 3 (EV_ABS), code 1 (ABS_Y), value 601594
Event: time 1698250985.875094, type 4 (EV_MSC), code 0 (MSC_SERIAL), value 1551894234
Event: time 1698250985.875094, -------------- SYN_REPORT ------------
Event: time 1698250985.879813, type 3 (EV_ABS), code 26 (ABS_TILT_X), value 5
Event: time 1698250985.879813, type 3 (EV_ABS), code 27 (ABS_TILT_Y), value -16
Event: time 1698250985.879813, type 3 (EV_ABS), code 0 (ABS_X), value 1260679
Event: time 1698250985.879813, type 3 (EV_ABS), code 1 (ABS_Y), value 597372
Event: time 1698250985.879813, type 4 (EV_MSC), code 0 (MSC_SERIAL), value 1551894234
Event: time 1698250985.879813, -------------- SYN_REPORT ------------
Event: time 1698250985.884813, type 3 (EV_ABS), code 26 (ABS_TILT_X), value 4
Event: time 1698250985.884813, type 3 (EV_ABS), code 27 (ABS_TILT_Y), value -15
Event: time 1698250985.884813, type 3 (EV_ABS), code 0 (ABS_X), value 1258641
Event: time 1698250985.884813, type 3 (EV_ABS), code 1 (ABS_Y), value 593326
Event: time 1698250985.884813, type 4 (EV_MSC), code 0 (MSC_SERIAL), value 1551894234
Event: time 1698250985.884813, -------------- SYN_REPORT ------------
Event: time 1698250985.889922, type 3 (EV_ABS), code 27 (ABS_TILT_Y), value -13
Event: time 1698250985.889922, type 3 (EV_ABS), code 0 (ABS_X), value 1253651
Event: time 1698250985.889922, type 3 (EV_ABS), code 1 (ABS_Y), value 588956
Event: time 1698250985.889922, type 4 (EV_MSC), code 0 (MSC_SERIAL), value 1551894234
Event: time 1698250985.889922, -------------- SYN_REPORT ------------
Event: time 1698250985.894922, type 3 (EV_ABS), code 27 (ABS_TILT_Y), value -12
Event: time 1698250985.894922, type 3 (EV_ABS), code 0 (ABS_X), value 1246771
Event: time 1698250985.894922, type 3 (EV_ABS), code 1 (ABS_Y), value 587480
Event: time 1698250985.894922, type 4 (EV_MSC), code 0 (MSC_SERIAL), value 1551894234
Event: time 1698250985.894922, -------------- SYN_REPORT ------------
Event: time 1698250985.899809, type 4 (EV_MSC), code 0 (MSC_SERIAL), value 1551894234
Event: time 1698250985.899809, -------------- SYN_REPORT ------------
Event: time 1698250985.913810, type 4 (EV_MSC), code 0 (MSC_SERIAL), value 1551894234
Event: time 1698250985.913810, -------------- SYN_REPORT ------------
Event: time 1698250985.922844, type 4 (EV_MSC), code 0 (MSC_SERIAL), value 1551894234
Event: time 1698250985.922844, -------------- SYN_REPORT ------------
Event: time 1698250985.931844, type 4 (EV_MSC), code 0 (MSC_SERIAL), value 1551894234
Event: time 1698250985.931844, -------------- SYN_REPORT ------------
Event: time 1698250985.941812, type 4 (EV_MSC), code 0 (MSC_SERIAL), value 1551894234
Event: time 1698250985.941812, -------------- SYN_REPORT ------------
Event: time 1698250985.950844, type 4 (EV_MSC), code 0 (MSC_SERIAL), value 1551894234
Event: time 1698250985.950844, -------------- SYN_REPORT ------------
Event: time 1698250985.959842, type 4 (EV_MSC), code 0 (MSC_SERIAL), value 1551894234
Event: time 1698250985.959842, -------------- SYN_REPORT ------------
Event: time 1698250985.969810, type 4 (EV_MSC), code 0 (MSC_SERIAL), value 1551894234
Event: time 1698250985.969810, -------------- SYN_REPORT ------------
Event: time 1698250985.978810, type 4 (EV_MSC), code 0 (MSC_SERIAL), value 1551894234
Event: time 1698250985.978810, -------------- SYN_REPORT ------------
Event: time 1698250985.987810, type 4 (EV_MSC), code 0 (MSC_SERIAL), value 1551894234
Event: time 1698250985.987810, -------------- SYN_REPORT ------------
Event: time 1698250985.997851, type 4 (EV_MSC), code 0 (MSC_SERIAL), value 1551894234
Event: time 1698250985.997851, -------------- SYN_REPORT ------------
Event: time 1698250986.008122, type 1 (EV_KEY), code 320 (BTN_TOOL_PEN), value 0
Event: time 1698250986.008122, type 3 (EV_ABS), code 40 (ABS_MISC), value 0
Event: time 1698250986.008122, type 4 (EV_MSC), code 0 (MSC_SERIAL), value 1551894234
Event: time 1698250986.008122, -------------- SYN_REPORT ------------
```

equivalent evtest:
```
libinput debug-events /dev/input/event11
-event11  DEVICE_ADDED            OpenTabletDriver Virtual Artist Tablet seat0 default group1  cap:T  size 45x14mm calib
 event11  TABLET_TOOL_PROXIMITY   +0.006s		12.54*/6.06*	tilt: 8.04*/-12.06*	pressure: 0.20*	pen      (0x5c8006da, id 0x10802) proximity-in 	axes:pt	btn:SS2
 event11  TABLET_TOOL_TIP         +0.006s		12.54/6.06	tilt: 8.04/-12.06	pressure: 0.20 down
 event11  TABLET_TOOL_TIP         +0.010s		12.58*/6.02*	tilt: 6.03*/-14.07*	pressure: 0.00* up
 event11  TABLET_TOOL_AXIS        +0.014s		12.61*/5.97*	tilt: 5.03*/-16.08*	pressure: 0.00
 event11  TABLET_TOOL_AXIS        +0.019s		12.59*/5.93*	tilt: 4.02*/-15.08*	pressure: 0.00
 event11  TABLET_TOOL_AXIS        +0.024s		12.54*/5.89*	tilt: 4.02/-13.07*	pressure: 0.00
 event11  TABLET_TOOL_AXIS        +0.029s		12.47*/5.87*	tilt: 4.02/-12.06*	pressure: 0.00
 event11  TABLET_TOOL_PROXIMITY   +0.143s		12.47/5.87	tilt: 4.02/-12.06	pressure: 0.00	pen      (0x5c8006da, id 0x10802) proximity-out 
 ```
 
I've had to implement an `IConfidenceHandler` for artist mode because otherwise tools do not report their ID and serial before when at least this return true, which confuses libinput if we don't halt input before the confidence threshold is met. This has the lovely return of the side effect that is reducing lift-off height, so this would potentially be nice to have as an option in the interface but I fear this requires #2254 before we can allow the user to disable this behavior, at the cost of compromising tool identification.

I will edit the OP to explain more in-depth, but I think this PR is a lot more ready for review and testing now.

**Comment by gonX** [comment]
I pushed another commit to increase the lift off height once the tool has been detected. It does not extend lift off height for tablets that do not have tool reporting.

File: OpenTabletDriver/Platform/Pointer/IToolHandler.cs [hunk 2]
```diff
@@ -0,0 +1,7 @@
+namespace OpenTabletDriver.Platform.Pointer
+{
+    public interface IToolHandler
```
Comment (gonX) [inline OpenTabletDriver/Platform/Pointer/IToolHandler.cs:3]: this needs xmldoc and likely [PublicAPI] like `IConfidenceHandler`

**Comment by gonX** [comment]
This needs to be retargeted to 0.6, and should have an UI element that can toggle this behavior with it being off by default, to avoid regressing the max distance the tool can start sending.
